### DISCUSSION TITLE
Bind expression to member of containing type's field

### DISCUSSION THREAD

**Discussion by joelpryde** [dotnet/roslyn#63971] [Q&A]
I'm trying to add an experimental compiler feature (for fun and profit) that allows the user to bind an expression directly to a member (field or method in this case) of an object contained in a type's field.  For instance, something like the following:

```csharp
public struct Position
{
    public float X;
}

public struct C 
{
    using Position position;
    
    public void M() 
    {
        // X here is bound to position.X
        X = 1.0f;
    }
}
```

My initial thought was to lex/parse the special case for `using` here and then to add a binding step to bind `X` in the `M()` method to a field provided by `position` (since this seems somewhat similar to how `using static` works - but for members of static types).  I've run into the issue however, that I can't figure out how to bind a single expression to a FieldSymbol or MethodSymbol that is contained in a specific object (in this case, my containing type's position field).  Is there a way to express this in the Roslyn binding stage - and any example therein - or is it a limitation of binding that it only allows binding individual expressions directly to accessible members of the containing type, locals and statics?

Obviously this is a bit beyond typical Roslyn usage, but I would appreciate any assistance.

**Comment by 333fred** [comment]
You would need to insert a binder in the binding chain to participate in LookupSymbols.

**Reply by joelpryde** [reply]
Also, I may have misunderstood.  Is it possible to modularly change how binding occurs for a specific piece of syntax (say one with a given modifier)?  That would be much cleaner than hacking into LookupSymbols/SynthesizeReceiver.

**Reply by 333fred** [reply]
> Also, I may have misunderstood. Is it possible to modularly change how binding occurs for a specific piece of syntax (say one with a given modifier)?

Sure, if you're modifying the compiler you're modifying the compiler.

> That would be much cleaner than hacking into LookupSymbols/SynthesizeReceiver.

What I'm suggesting isn't a hack, it's the correct way of resolving the issue.

**Reply by 333fred** [reply]
Just to clarify: I'm talking about how to bind the `X`, not the field. The field will likely need to have parsing changes so you can make it appear as a field. Once it's a field, you can add it and all other such fields to a new binder type in the chain when binding method bodies.

**Reply by joelpryde** [reply]
I think there was a misunderstanding.  I was referring to the way I had implemented as a hack - not what you are suggesting.

I'll try adding a new binder.

**Comment by joelpryde** [comment]
We ended up trying that.  Unfortunately LookupResult just contains a list of individual symbols (whereas we need to bind our original expression to a field lookup inside another field - thus two, or more, FieldSymbols).

What ended up working was looking up our leaf FieldSymbol in LookupSymbols and then later constructing a set of BoundThisReference and BindFieldAccess nodes for the bound tree when we synthesized the receiver for the field access.  Our eventual bound operation tree for our expression looked something like this:

```
IFieldReferenceOperation: System.Int32 A.X (OperationKind.FieldReference, Type: System.Int32) (Syntax: 'X')
    Instance Receiver:
        IFieldReferenceOperation: A C.position (OperationKind.FieldReference, Type: Position, IsImplicit) (Syntax: 'X')
            Instance Receiver:
                IInstanceReferenceOperation (ReferenceKind: ContainingTypeInstance) (OperationKind.InstanceReference, Type: C, IsImplicit) (Syntax: 'X')
```

with the middle FieldReferenceOperation representing the additional piece of information for referencing our implicitly referenced field in our original type.

**Reply by 333fred** [reply]
> Unfortunately LookupResult just contains a list of individual symbols

It sounds like you're not inserting a _new_ binder, but trying to reuse an existing one. You need to implement a new derived binder type and insert it into the correct point in the binder chain, so lookup can occur correctly.
